db/inlineskiplist.h

```diff
@@ -558,34 +696,191 @@ void InlineSkipList<Comparator>::Insert(const char* key) {
+  InsertHint* hint = reinterpret_cast<InsertHint*>(*hint_ptr);
+  if (hint == nullptr) {
+    // AllocateInsertHint will initialize hint with num_levels = 0 and
+    // prev[i] = head_ for all i.
+    hint = AllocateInsertHint();
+    *hint_ptr = reinterpret_cast<void*>(hint);
   }
 
-  for (int i = 0; i < height; i++) {
-    // NoBarrier_SetNext() suffices since we will add a barrier when
-    // we publish a pointer to "x" in prev[i].
-    x->NoBarrier_SetNext(i, prev_[i]->NoBarrier_Next(i));
-    prev_[i]->SetNext(i, x);
+  // Look for the first level i < num_levels with prev[i] < key.
+  int level = 0;
+  for (; level < hint->num_levels; level++) {
+    if (KeyIsAfterNode(key, hint->prev[level])) {
+      assert(!KeyIsAfterNode(key, hint->prev[level]->Next(level)));
+      break;
+    }
+  }
+  // Update prev above num_levels level if needed.
+  if (level >= hint->num_levels) {
+    // We need to find a node on level equal to "level" to search for insert
+    // position, and we need to insert the new node to every level below
+    // "height", so we need to update prev for all levels in
+    // [level, max(level, height-1)].
+    FindWithHint(key, hint, std::max<int>(level, height - 1), level);
+  } else {
+    // Get the max height among prev[level]...prev[num_level-1]. For levels in
+    // [level, max_prev_height) the new node will be inserted after one of
+    // prev[i]. We only need to update prev on and above max_prev_height.
+    int max_prev_height = 0;
+    for (int i = level; i < hint->num_levels; i++) {
+      max_prev_height = std::max<int>(max_prev_height, hint->prev_height[i]);
+    }
+    if (height > max_prev_height) {
+      FindWithHint(key, hint, height - 1, max_prev_height);
+    }
+  }
+  // The current level where the new node is to insert into skip-list.
+  int current_level = 0;
+  // For each level below "level", search for the insert position and insert
+  // the node.
+  if (level > 0) {
+    Node* tmp[level];
+    FindLessThan(key, tmp, hint->prev[level], level, 0);
+    while (current_level < height && current_level < level) {
+      assert(KeyIsAfterNode(key, tmp[current_level]));
+      assert(!KeyIsAfterNode(key, tmp[current_level]->Next(current_level)));
+      x->InsertAfter(tmp[current_level], current_level);
+      current_level++;
+    }
+  }
+  // For each level h in [level, max_prev_height), there is a level i in
+  // [level, num_levels) where
+  //   * prev_height[i] > h.
+  //   * prev_height[j] <= h for all level < j < i.
+  // Since there's no node in the range of (prev[i], prev[level]) has height
+  // greater than h, the new node can be inserted after prev[i] on level h.
+  for (int i = level; i < hint->num_levels; i++) {
+    while (current_level < height && current_level < hint->prev_height[i]) {
+      assert(KeyIsAfterNode(key, hint->prev[i]));
+      assert(!KeyIsAfterNode(key, hint->prev[i]->Next(current_level)));
+      x->InsertAfter(hint->prev[i], current_level);
```

Comment:
One thing made it hard for me to understand the code is the dual function of hint->prev. It is used as the location to insert the key here, but some part of it is also the position to insert the new entry.

If we can separate the two. Use another local array for the position to insert, just as `tmp` in line 828 to 835, it may be easier to understand.